rust_src/src/buffers.rs

```diff
@@ -1982,4 +2030,268 @@ pub fn byte_char_debug_check(b: LispBufferRef, charpos: isize, bytepos: isize) {
     }
 }
 
+/// Kill the buffer specified by BUFFER-OR-NAME.
+/// The argument may be a buffer or the name of an existing buffer.
+/// Argument nil or omitted means kill the current buffer.  Return t if the
+/// buffer is actually killed, nil otherwise.
+///
+/// The functions in `kill-buffer-query-functions' are called with the
+/// buffer to be killed as the current buffer.  If any of them returns nil,
+/// the buffer is not killed.  The hook `kill-buffer-hook' is run before the
+/// buffer is actually killed.  The buffer being killed will be current
+/// while the hook is running.  Functions called by any of these hooks are
+/// supposed to not change the current buffer.
+///
+/// Any processes that have this buffer as the `process-buffer' are killed
+/// with SIGHUP.  This function calls `replace-buffer-in-windows' for
+/// cleaning up all windows currently displaying the buffer to be killed.
+#[lisp_fn(min = "0", intspec = "bKill buffer: ")]
+pub fn kill_buffer(buffer_or_name: Option<LispBufferOrName>) -> bool {
+    let mut b = match buffer_or_name {
+        Some(buf_or_name) => LispBufferRef::from(buf_or_name),
+        None => ThreadState::current_buffer_unchecked(),
+    };
+
+    // Avoid trouble for buffers that are already dead.
+    if !b.is_live() || unsafe { thread_check_current_buffer(b.as_mut()) } {
+        return false;
+    }
+
+    // Run hooks with the buffer to be killed the current buffer.
+    let count = unsafe { SPECPDL_INDEX() };
+
+    unsafe {
+        record_unwind_protect(Some(save_excursion_restore), save_excursion_save());
+        set_buffer_internal(b.as_mut());
+    }
+
+    // First run the query functions; if any query is answered no,
+    // don't kill the buffer.
+    if !run_hook_with_args_until_failure(&mut [Qkill_buffer_query_functions]) {
+        return unbind_to(count, Qnil).into();
+    }
+
+    // Query if the buffer is still modified.
+    if is_interactive() && b.filename().is_not_nil() && b.modified_since_save() {
+        // TODO: This was AUTO_STRING.
+        let format_str = new_unibyte_string!("Buffer %s modified; kill anyway? ");
+        if !yes_or_no_p(format(&mut [format_str, b.name_]).into()) {
+            return unbind_to(count, Qnil).into();
+        }
+    }
+
+    // If the hooks have killed the buffer, exit now.
+    if !b.is_live() {
+        return unbind_to(count, Qt).into();
+    }
+
+    // Then run the hooks.
+    unsafe { run_hook(Qkill_buffer_hook) };
+    unbind_to(count, Qnil);
+
+    // If the hooks have killed the buffer, exit now.
+    if !b.is_live() {
+        return true;
+    }
+
+    // We have no more questions to ask.  Verify that it is valid
+    // to kill the buffer.  This must be done after the questions
+    // since anything can happen within do_yes_or_no_p.
+
+    // Don't kill the minibuffer now current.
+    if LispObject::from(b) == LispWindowRef::from(unsafe { minibuf_window }).contents {
```

Comment:
This one reads odd. On the left we have a LispObject and on the right we have a LispWindowRef which is not a LispObject.

Reply:
It's a byproduct of having to access a `LispObject` member of the `minibuf_window`, which is also stored as a `LispObject`. I've added a function for accessing `minibuf` that converts it to a `LispWindowRef` in my local tree, which makes this look far more readable.